---
title: "SeqAnalysis"
author: "Erik Killian"
date: "2024-03-28"
output: html_document
editor_options: 
  chunk_output_type: console
---
## Setup
```{r setup, include=FALSE}
setwd("Projects/MicrobeAnalysis")

library(ggplot2)
library(knitr)
library(gridExtra)
library(dada2)
library(phyloseq)
library(microbiome)
library(vegan)
library(dplyr)
library(plyr)
library(tidyverse)
library(file2meco)
library(ggordiplots)
library(randomForest)
library(microeco)
library(ggh4x)
library("ape")

```

```{r creating psField (one and done), eval=FALSE}
ps2021 = readRDS("Projects/2021_Field/Bacteria/2021Phyloseq.RDS")
ps2022 = readRDS("Projects/2022_Field/2022Phyloseq.RDS")

psField = merge_phyloseq(ps2021,ps2022)

random_tree = rtree(ntaxa(psField), rooted=TRUE, tip.label=taxa_names(psField))
psField <- merge_phyloseq(psField, random_tree)


# Cleaning sample data
x = sample_data(psField)
x$Line[grepl("B", x$Plot)] = "Bulk"
class(x) = "data.frame"
str(x)
y = merge(x, HFA, by.x = "Line", by.y = "Line", all.x = TRUE)
y = subset(y, select = -c(HFA))
colnames(y)[colnames(y) == "Location.y"] = "HomeSite" # Homesite defined through HFA calculation using 2021-2022 site years. 
BP = y[!is.na(y$breeding_program),]
BP = BP[,c("Line","breeding_program")]
BP = unique(BP)
y = merge(y,BP, by = "Line", all.x = TRUE)
y = subset(y, select = -c(breeding_program.x))
colnames(y)[colnames(y) == "breeding_program.y"] = "BreedingProgram"
colnames(y)[colnames(y) == "Location.x"] = "Location"
row.names(y) = paste(y$Year,y$Location,y$Plot, sep = "_")

sample_data(psField) = x


```

## CHAPTER 1: BASIC ANALYSES

```{r Alpha Diversity, message=FALSE,warning=FALSE, fig.height=5,fig.width=10}

a = microbiome::alpha(psField, index = "Observed")
samdata = as.data.frame(sample_data(psField))
alphaStat = merge(samdata,a, by = "row.names")

balphaStat = alphaStat %>%
  filter(grepl("B", Plot))

alphaStat <- subset(alphaStat, !grepl("B",Plot))

###
#PLOTS 
###
ggplot(data = alphaStat[!is.na(alphaStat$Block),], aes(x = factor(Block, levels = c(1,2,3,4,5,6,7,8,9,10,11,12)), y = observed, fill = Location)) + geom_boxplot() + facet_grid(~ Year, scales = "free_x") +  
  xlab("Block") + 
  ylab("Observed Diversity") + 
  ggtitle("Yearly Observed Diversity by Plot") + 
  theme_classic() + 
  theme(
    text = element_text(family = "serif")
  )

ggplot(data = alphaStat[!is.na(alphaStat$Block),], aes(x = Location, y = observed, fill = Year)) + geom_boxplot() + 
  ylab("Observed Diversity") + 
  ggtitle("ASV Diversity in Rhizosphere by Location-Year") + 
  scale_fill_discrete(type = c("peachpuff3","steelblue3")) + 
  theme_classic()

ggplot(data = balphaStat, aes(x = Location, y = observed, fill = Year)) + geom_boxplot() + 
  ylab("Observed Diversity") + 
  ggtitle("ASV Diversity in Bulk Soil by Location-Year") + 
  scale_fill_discrete(type = c("lightsalmon2","mediumpurple2")) + 
  theme_classic()

```

```{r Relative Abundance}
table(tax_table(psField)[, "Phylum"], exclude = NULL)
ps1 <- subset_taxa(psField, !is.na(Kingdom) & !Kingdom %in% "k__Archaea")
ps1 <- subset_taxa(ps1, !is.na(Phylum) & !Phylum %in% c("", "uncharacterized", "p__FBP", "p__Fibrobacteres", "p__Cyanobacteria", "Chlorophyta_ph"))
ps1
#saveRDS(ps1, "ps1.RDS")
#ps1 = readRDS("ps1.RDS")


pst <- ps1 %>%
 tax_glom(taxrank = "Phylum") %>% # agglomerate at phylum level
 transform_sample_counts(function(x) {x/sum(x)} ) # Transform to rel. abundance

psf <- pst %>% psmelt() %>% # Melt to long format
 #filter(Abundance > 0.02) %>% # Filter out low abundance taxa
 arrange(Phylum) # Sort data frame alphabetically by phylum

psf = psf[,c("Sample","Abundance","Phylum")]
psf$Year = substr(psf$Sample,1,4)
psf$Location = substr(psf$Sample, 6,7)
psf$Sample = substr(x = psf$Sample, 9, length(psf$Sample))
#saveRDS(psf,"psf.RDS")

bulk = psf %>%
  filter(grepl("B", Sample))
```

```{r Relative Abundance Plotting, fig.width=8,fig.height=7}
# Calculate mean abundance across locations! Why did I do that...
psm_summarized <- ddply(psf, c("Sample","Phylum"), summarise, N = length(Abundance), mean = mean(Abundance))
#saveRDS(psm_summarized, "psm_summarized.RDS")
View(psm_summarized)

# Set colors for plotting
phylum_colors <- c(
 "#CBD588", "#5F7FC7", "orange","#DA5724", "#508578", "#CD9BCD",
 "#AD6F3B", "#673770","#D14285", "#652926", "#C84248", 
 "#8569D5", "#5E738F","#D1A33D", "#8A7C64", "#599861"
)

myColors = c("Acidobacteriota" = "#CBD588", "Actinobacteriota" = "#5F7FC7", "Bacteroidota" = "orange", "Chloroflexi" = "#508578", "Firmicutes" = "#DA5724", "Gemmatimonadota" = "#CD9BCD", "Myxococcota" = "#AD6F3B", "Planctomycetota" = "#673770", "Proteobacteria" = "#D14285", "Verrucomicrobiota" = "#652926")

myPhyla = c("Verrucomicrobiota","Myxococcota","Planctomycetota","Gemmatimonadota","Firmicutes","Acidobacteriota","Bacteroidota", "Chloroflexi", "Proteobacteria", "Actinobacteriota")

cc = scale_colour_manual(name = myPhyla, values = myColors)

psf.rhiz = filter(psf, !grepl("B",Sample))
psf.bulk = filter(psf, grepl("B",Sample))
########################
# Plot Relative Abundance of Everything other than the Checks
########################
ggplot(psf.rhiz, aes(x = Sample, y = Abundance, fill = factor(Phylum, levels = myPhyla))) + facet_grid(Year ~ Location,scales = "free") +
  geom_bar(stat = "identity", width = 1) +
  ggtitle("Relative Abundance of Microbial Phyla in Barley Rhizosphere") +
  xlab("Barley Variety") + ylab("Relative Abundance") + 
  scale_y_continuous(breaks = c(0.25,0.5,0.75,1),expand = c(0,0,0,0.0)) +
  scale_fill_manual(values = myColors, name = "Phylum") +
  theme_classic() + 
    theme(
      text = element_text(size = 16, family = "serif"),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line.y = element_blank(),
      plot.title = element_text(hjust = 0, size = 18),
      legend.title = element_text(hjust = 0.5, size = 16)
      ) 

ggplot(psf.bulk, aes(x = Sample, y = Abundance, fill = factor(Phylum, levels = myPhyla))) + facet_grid(Year ~ Location,scales = "free") +
  geom_bar(stat = "identity", width = 1) +
  ggtitle("Relative Abundance of Microbial Phyla in Bulk Soil") +
  xlab("Barley Variety") + ylab("Relative Abundance") + 
  scale_y_continuous(breaks = c(0.25,0.5,0.75,1),expand = c(0,0,0,0.0)) +
  scale_fill_manual(values = myColors, name = "Phylum") +
  theme_classic() + 
    theme(
      text = element_text(size = 16, family = "serif"),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line.y = element_blank(),
      plot.title = element_text(hjust = 0, size = 18),
      legend.title = element_text(hjust = 0.5, size = 16)
      ) 


#############################
# Reorganizing plot for paper
#############################
View(psf)
psf.rhiz = filter(psf, !grepl("B",Sample))
x = psf.rhiz[psf.rhiz$Location == "BZ"& psf.rhiz$Year == "2021",]
x = na.omit(x)

g0 = ggplot(psf.rhiz[psf.rhiz$Location == "BZ"& psf.rhiz$Year == "2021",], aes(x = Sample, y = Abundance,fill = factor(Phylum, levels = myPhyla))) + geom_bar(stat = "identity",width = 1) +
    ggtitle("Bozeman 2021") +
  xlab("") + ylab("") + 
  scale_y_continuous(breaks = c(0.25,0.5,0.75,1),expand = c(0,0,0,0.0)) +
  scale_fill_manual(values = myColors, name = "Phylum") +
  theme_classic() + 
    theme(
      text = element_text(size = 16),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line.y = element_blank(),
      plot.title = element_text(hjust = 0, size = 16),
      legend.title = element_text("Phylum")
      ) 

g1 = ggplot(psf.rhiz[psf.rhiz$Location == "BZ"& psf.rhiz$Year == "2021",], aes(x = Sample, y = Abundance,fill = factor(Phylum, levels = myPhyla))) + geom_bar(stat = "identity",width = 1) +
    ggtitle("Bozeman 2021") +
  xlab("") + ylab("") + 
  scale_y_continuous(breaks = c(0.25,0.5,0.75,1),expand = c(0,0,0,0.0)) +
  scale_fill_manual(values = myColors, name = "Phylum") +
  theme_classic() + 
    theme(
      text = element_text(size = 14),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line.y = element_blank(),
      plot.title = element_text(hjust = 0, size = 16),
      legend.position = "none"
      ) 

g2 = ggplot(psf.rhiz[psf.rhiz$Location == "BZ"& psf.rhiz$Year == "2022",], aes(x = Sample, y = Abundance,fill = factor(Phylum, levels = myPhyla))) + geom_bar(stat = "identity",width = 1) +
    ggtitle("Bozeman 2022") +
  xlab("") + ylab("") + 
  scale_y_continuous(breaks = c(0.25,0.5,0.75,1),expand = c(0,0,0,0.0)) +
  scale_fill_manual(values = myColors, name = "Phylum") +
  theme_classic() + 
    theme(
      text = element_text(size = 14),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line.y = element_blank(),
      plot.title = element_text(hjust = 0, size = 16),
      legend.position = "none"
      ) 

g3 = ggplot(psf.rhiz[psf.rhiz$Location == "SD"& psf.rhiz$Year == "2021",], aes(x = Sample, y = Abundance,fill = factor(Phylum, levels = myPhyla))) + geom_bar(stat = "identity",width = 1) +
    ggtitle("South Dakota 2021") +
  xlab("") + ylab("") + 
  scale_y_continuous(breaks = c(0.25,0.5,0.75,1),expand = c(0,0,0,0.0)) +
  scale_fill_manual(values = myColors, name = "Phylum") +
  theme_classic() + 
    theme(
      text = element_text(size = 14),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line.y = element_blank(),
      plot.title = element_text(hjust = 0, size = 16),
      legend.position = "none"
      ) 

g4 = ggplot(psf.rhiz[psf.rhiz$Location == "SD"& psf.rhiz$Year == "2022",], aes(x = Sample, y = Abundance,fill = factor(Phylum, levels = myPhyla))) + geom_bar(stat = "identity",width = 1) +
    ggtitle("South Dakota 2022") +
  xlab("") + ylab("") + 
  scale_y_continuous(breaks = c(0.25,0.5,0.75,1),expand = c(0,0,0,0.0)) +
  scale_fill_manual(values = myColors, name = "Phylum") +
  theme_classic() + 
    theme(
      text = element_text(size = 14),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line.y = element_blank(),
      plot.title = element_text(hjust = 0, size = 16),
      legend.position = "none"
      ) 

g5 = ggplot(psf.rhiz[psf.rhiz$Location == "MO"& psf.rhiz$Year == "2021",], aes(x = Sample, y = Abundance,fill = factor(Phylum, levels = myPhyla))) + geom_bar(stat = "identity",width = 1) +
    ggtitle("CARC 2021") +
  xlab("") + ylab("") + 
  scale_y_continuous(breaks = c(0.25,0.5,0.75,1),expand = c(0,0,0,0.0)) +
  scale_fill_manual(values = myColors, name = "Phylum") +
  theme_classic() + 
    theme(
      text = element_text(size = 14),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line.y = element_blank(),
      plot.title = element_text(hjust = 0, size = 16),
      legend.position = "none"
      )

g6 = ggplot(psf.rhiz[psf.rhiz$Location == "MO"& psf.rhiz$Year == "2022",], aes(x = Sample, y = Abundance,fill = factor(Phylum, levels = myPhyla))) + geom_bar(stat = "identity",width = 1) +
    ggtitle("CARC 2022") +
  xlab("") + ylab("") + 
  scale_y_continuous(breaks = c(0.25,0.5,0.75,1),expand = c(0,0,0,0.0)) +
  scale_fill_manual(values = myColors, name = "Phylum") +
  theme_classic() + 
    theme(
      text = element_text(size = 14),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line.y = element_blank(),
      plot.title = element_text(hjust = 0, size = 16),
      legend.position = "none"
      )

g7 = ggplot(psf.rhiz[psf.rhiz$Location == "HI"& psf.rhiz$Year == "2021",], aes(x = Sample, y = Abundance,fill = factor(Phylum, levels = myPhyla))) + geom_bar(stat = "identity",width = 1) +
    ggtitle("Hawaii 2021") +
  xlab("") + ylab("") + 
  scale_y_continuous(breaks = c(0.25,0.5,0.75,1),expand = c(0,0,0,0.0)) +
  scale_fill_manual(values = myColors, name = "Phylum") +
  theme_classic() + 
    theme(
      text = element_text(size = 14),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line.y = element_blank(),
      plot.title = element_text(hjust = 0, size = 16),
      legend.position = "none"
      )

g0
grid.arrange(g1,g5,g3,g2,g6,g4,g7) # Exported to powerpoint to make it look good.
```

```{r Bulk Relative Abundance Plots}
#########################
# Bulk Relative Abundances
##########################
x = psf.bulk[psf.bulk$Location == "BZ"& psf.bulk$Year == "2021",]
b1 = ggplot(psf.bulk[psf.bulk$Location == "BZ"& psf.bulk$Year == "2021",], aes(x = Sample, y = Abundance,fill = factor(Phylum, levels = myPhyla))) + geom_bar(stat = "identity",width = 1) +
    ggtitle("Bozeman 2021") +
  xlab("") + ylab("") + 
  scale_y_continuous(breaks = c(0.25,0.5,0.75,1),expand = c(0,0,0,0.0)) +
  scale_fill_manual(values = myColors, name = "Phylum") +
  theme_classic() + 
    theme(
      text = element_text(size = 14),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line.y = element_blank(),
      plot.title = element_text(hjust = 0, size = 16),
      legend.position = "none"
      )

b2 = ggplot(psf.bulk[psf.bulk$Location == "BZ"& psf.bulk$Year == "2022",], aes(x = Sample, y = Abundance,fill = factor(Phylum, levels = myPhyla))) + geom_bar(stat = "identity",width = 1) +
    ggtitle("Bozeman 2022") +
  xlab("") + ylab("") + 
  scale_y_continuous(breaks = c(0.25,0.5,0.75,1),expand = c(0,0,0,0.0)) +
  scale_fill_manual(values = myColors, name = "Phylum") +
  theme_classic() + 
    theme(
      text = element_text(size = 14),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line.y = element_blank(),
      plot.title = element_text(hjust = 0, size = 16),
      legend.position = "none"
      )

b4 = ggplot(psf.bulk[psf.bulk$Location == "MO"& psf.bulk$Year == "2022",], aes(x = Sample, y = Abundance,fill = factor(Phylum, levels = myPhyla))) + geom_bar(stat = "identity",width = 1) +
    ggtitle("CARC 2022 Pooled") +
  xlab("") + ylab("") + 
  scale_y_continuous(breaks = c(0.25,0.5,0.75,1),expand = c(0,0,0,0.0)) +
  scale_fill_manual(values = myColors, name = "Phylum") +
  theme_classic() + 
    theme(
      text = element_text(size = 14),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line.y = element_blank(),
      plot.title = element_text(hjust = 0, size = 16),
      legend.position = "none"
      )

b5 = ggplot(psf.bulk[psf.bulk$Location == "SD"& psf.bulk$Year == "2021",], aes(x = Sample, y = Abundance,fill = factor(Phylum, levels = myPhyla))) + geom_bar(stat = "identity",width = 1) +
    ggtitle("South Dakota 2021") +
  xlab("") + ylab("") + 
  scale_y_continuous(breaks = c(0.25,0.5,0.75,1),expand = c(0,0,0,0.0)) +
  scale_fill_manual(values = myColors, name = "Phylum") +
  theme_classic() + 
    theme(
      text = element_text(size = 14),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line.y = element_blank(),
      plot.title = element_text(hjust = 0, size = 16),
      legend.position = "none"
      )

b6 = ggplot(psf.bulk[psf.bulk$Location == "SD"& psf.bulk$Year == "2022",], aes(x = Sample, y = Abundance,fill = factor(Phylum, levels = myPhyla))) + geom_bar(stat = "identity",width = 1) +
    ggtitle("South Dakota 2022") +
  xlab("") + ylab("") + 
  scale_y_continuous(breaks = c(0.25,0.5,0.75,1),expand = c(0,0,0,0.0)) +
  scale_fill_manual(values = myColors, name = "Phylum") +
  theme_classic() + 
    theme(
      text = element_text(size = 14),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line.y = element_blank(),
      plot.title = element_text(hjust = 0, size = 16),
      legend.position = "none"
      )

grid.arrange(b1,b2,b4,b5,b6)
```

This can replace all the above "plotting relative abundance chunks." 
```{r Relative Abundance Plotting with MicroEco}

me.psField = phyloseq2meco(psField)
me.psField$sample_table = subset(me.psField$sample_table, Line != "Bulk")

relAbun <- trans_abund$new(dataset = me.psField, taxrank = "Phylum", ntaxa = 8)
relAbun$data_abund$Location = factor(relAbun$data_abund$Location, levels = c("BZ","MO","SD","HI"))

relAbun$plot_bar(others_color = "grey70", facet = c("Location","Year"), xtext_keep = FALSE, legend_text_italic = FALSE, barwidth = 1)


```



## CHAPTER 2: ORDINATION

First I calculated weighted unifrac across all location-years. Next I calculated weighted unifrac for only the mainland locations (BZ,MO,SD). 

To calculate the effects of factors, I ran a permanova on the weighted unifrac distance matrix. The original, using ordinate() contains positive and negative values, which doesn't work with permanovas. Therefore I re-ran the weighted unifrac with UniFrac() and "normalized" = TRUE. 
```{r Ordination, fig.height=5,fig.width=6}
rhizo.ps = subset_samples(ps1, Line != "Bulk")


## RDA Updated Way ------------
uni = UniFrac(rhizo.ps, weighted = TRUE, normalized = TRUE, fast = TRUE) # Same as ordinate, but I'm normalizing for permanova

sample = as.data.frame(as.matrix(sample_data(rhizo.ps)))
sample$locyr = paste(sample$Location, sample$Year, sep = "_")

# Model by Location and Year *This is the one*
r2 = capscale(uni ~ Location : Year, data = sample)
p2 = gg_ordiplot(r2, groups = sample$Location)
p2$plot + scale_color_manual(name = "Location",
                     labels = c("swMT","HI","cntrMT","SD"),
                     values = c("skyblue","goldenrod1","darkorange","seagreen3")) + theme_classic()
summary(r2)
anova(r2)
screeplot(r2)
anova(r2, by = "term")

# Model includes Year, Location, and interaction
r3 = capscale(uni ~ Location + Year + Location:Year, data = sample)
p2 = gg_ordiplot(r2, groups = sample$Location)

anova(r3, by = "term")


# Model only by location
r = capscale(uni ~ Location,data = sample) # response is wunifrac distance matrix
p = gg_ordiplot(r, groups = sample$Location)
p$plot +
  scale_color_manual(name = "Location",
                     labels = c("swMT","HI","cntrMT","SD"),
                     values = c("skyblue","goldenrod1","darkorange","seagreen3")) + 
  theme_classic()

summary(r)
# Model by location with year as a condition (random variable)
## This means abundances are normalized across year before modeling by location. 
r1 = capscale(uni ~ Location + Condition(Year:Location), data = sample)
p1 = gg_ordiplot(r1, groups = sample$Location)
p1$plot + theme_classic()
summary(r1)
anova(r1)
screeplot(r1)



# Model by Location and Year 
r3 = capscale(uni ~ Location + Year, data = sample)
p3 = gg_ordiplot(r2, groups = sample$Location)
p3$plot + theme_classic()
summary(r3)
anova(r3)
screeplot(r3)
anova(r3, by = "onedf")


# All data --------
N.ord = ordinate(ps1, "PCoA", "wunifrac")

#saveRDS(N.ord, "wuPCoA.RDS")
N1.ord = readRDS("wuPCoA.RDS")



sample_data(ps1)$Location <- factor(sample_data(ps1)$Location, levels = c("BZ", "MO", "SD", "HI"))
plot_ordination(ps1, N1.ord,type = "sites", color = "Location", shape = "Year", title = "Weighted Unifrac by Location by Year") +
  scale_color_manual(name = "Location",
                     labels = c("swMT","cntrMT","SD","HI"),
                     values = c("skyblue","darkorange","seagreen3","goldenrod1")) +
  theme_classic()

x = data.frame(sample_data(ps1))


# Ordination Stats

metadata = as(sample_data(rhizo.ps), "data.frame")
metadata$locyr = paste(metadata$Location, metadata$Year,sep = "")

```

## CHAPTER 3: RANDOM FOREST

**Use in GWAS paper Figure 1**
```{r Random Forest without Hawaii across all locations, fig.width=12,fig.height=5}
# Create microEco object from ps1
meF = phyloseq2meco(ps1)
meF$sample_table = subset(meF$sample_table, Line != "Bulk")
meF$sample_table = subset(meF$sample_table, Location != "HI")
meFg = meF$merge_taxa(taxa = "Genus")
meFg$cal_abund()
## Prep microeco dataframe for random forest if needed
# meFg$tax_table$Genus = ifelse(nchar(meFg$tax_table$Genus)>0,yes = meFg$tax_table$Genus,no = "NA") # make blanks = NA
# meFg$tax_table = subset(meFg$tax_table, Genus != "NA")
# meFg$tax_table$Genus[meFg$tax_table$Genus == "Allorhizobium-Neorhizobium-Pararhizobium-Rhizobium"] = "Allo-Neo-Para-Rhizobium"

t1 <- trans_diff$new(dataset = meFg, method = "rf", group = "Location", taxa_level = "Genus", )

p.2.2 = t1$plot_diff_bar(use_number = 1:20, width = 0.8,keep_prefix = FALSE, 
                  group_order = c("BZ","MO", "SD")) +
    scale_fill_manual(name = "Location", 
     labels = c("swMT","cntrMT","SD"),
     values = c("skyblue","darkorange","seagreen3")) +
  scale_color_manual(name = "Location",
     labels = c("swMT","cntrMT","SD"),
     values = c("skyblue","darkorange","seagreen3"))

p.2.1 = t1$plot_diff_abund(add_sig = F, select_taxa = t1$plot_diff_bar_taxa, keep_prefix = FALSE) + ggplot2::ylab("Relative abundance (%)") +
    scale_fill_manual(name = "Location", 
     labels = c("SD","cntrMT","swMT"),
     values = c("seagreen3","darkorange","skyblue")) +
  scale_color_manual(name = "Location",
     labels = c("SD","cntrMT","swMT"),
     values = c("seagreen3","darkorange","skyblue"))

grid.arrange(p.2.2 + theme(legend.position = "none"),p.2.1, nrow = 1)
```

```{r Random Forest by Function, fig.width=12,fig.height=5}
####
# RF Across all FAPROTAX functions
####
meF = phyloseq2meco(ps1)
meF$sample_table = subset(meF$sample_table, Line != "Bulk")
meF$sample_table = subset(meF$sample_table, Location != "HI")
meFg = meF$merge_taxa(taxa = "Genus")
# meFg$tax_table$Genus = ifelse(nchar(meFg$tax_table$Genus)>0,yes = meFg$tax_table$Genus,no = "NA") # make blanks = NA
# meFg$tax_table = subset(meFg$tax_table, Genus != "NA")
# meFg$tax_table$Genus[meFg$tax_table$Genus == "Allorhizobium-Neorhizobium-Pararhizobium-Rhizobium"] = "Allo-Neo-Para-Rhizobium"
meFg$cal_abund()
meFg$sample_table$LocYear = paste(meFg$sample_table$Year, meFg$sample_table$Location, sep = "_")




func = trans_func$new(meFg)
func$cal_spe_func(prok_database = "FAPROTAX") # Run function assignment

# calculate the percentages for communities
# here consider the abundance
func$cal_spe_func_perc(abundance_weighted = TRUE)
func$trans_spe_func_perc()

meFg2 = clone(meFg)
# transpose res_spe_func_perc to be a data.frame like taxonomic abundance
tmp <- as.data.frame(t(func$res_spe_func_perc), check.names = FALSE)
# assign the table back to taxa_abund list for further analysis
meFg2$taxa_abund$func <- tmp
meFg2$sample_table$LocYear = paste(meFg2$sample_table$Year,meFg2$sample_table$Location, sep = "_")

# select the "func" in taxa_abund list in trans_diff
t2 <- trans_diff$new(dataset = meFg2, method = "rf", group = "Location", taxa_level = "func", importance = TRUE, boots = 30, filter_thres = 1)

p1 = t2$plot_diff_bar(width = 0.8, 
                  group_order = c("BZ","MO", "SD")) +
  scale_fill_manual(name = "Location", 
     labels = c("swMT","cntrMT","SD"),
     values = c("skyblue","darkorange","seagreen3")) +
  scale_color_manual(name = "Location",
     labels = c("swMT","cntrMT","SD"),
     values = c("skyblue","darkorange","seagreen3")) + 
  theme(legend.position = "none", axis.text.y = element_text(angle = 45, size = 9))
  
p2 = t2$plot_diff_abund(add_sig = F, select_taxa = t2$plot_diff_bar_taxa) + ggplot2::ylab("Relative abundance (%)") +
  scale_fill_manual(name = "Location", 
     labels = c("SD","cntrMT","swMT"),
     values = c("seagreen3","darkorange","skyblue")) +
  scale_color_manual(name = "Location",
     labels = c("SD","cntrMT","swMT"),
     values = c("seagreen3","darkorange","skyblue"))+theme(axis.text.y = element_blank())

p1 | p2


grid.arrange(p.2.4 + theme(legend.position = "none"), p.2.3,nrow = 1)

```
